Adapt to pkgconf 2.5.0

ppisar opened the pull request:
This patchset adapts to changes in pkgconf 1.9 and 2.5.0.

Upstream stopped adding trailing spaces to cflags/libs strings. This patch puts them back as that would break a use case when one concatenates various flags for a command line.

Most complex part is dealing with changed pkgconf dependency solver. pkgconf now returns only direct cflags/libs. One needs to use a pkgconf solver which populates an in-client cache and stores intermediate results into a dedicated PKGCONF_PKG_PROPF_VIRTUAL package.

Because the solver ignores packages loaded from a file, this patch also temporarily injects these packages into a cache to retrieve flags including dependencies.

All tests pass with pkgconf 1.9.4 and 1.8.0.

https://github.com/PerlAlien/PkgConfig-LibPkgConf/issues/15

ppisar committed:
Adapt to pkgconf-1.9.4

Fixes error handled prototype.
Adapts to passing flags to file_open.
Adapts tests to no traling spaces (incompatible change).
Works around a possible pkgconf regression with trailing null bytes.

TODO: t/package.t fails on missing shared libs from dependencies.
TODO: Restore compatibility with older pkgconfs.

ppisar committed:
Use solver for cflags/libs

Some t/package.t are fixed now.
But packages loaded with package_from_file() fails on the solver. See
t/package.t.

ppisar committed:
Cache packages loaded from files

Solver works on cached packages. Packages are cached either manually,
or by exploring all installation paths. To obtain cflags/libs with the
solver we need to cache the packages loaded from files.

Drawbacks:

(1) pkgconf tool flushed the cache before retrieving nonstatic libs
("reset solver when solving for library groups"). I don't understand
why. I was unable to observe any effect. Because we stash the package
into a cache, we cannot drop the cache when retrieving the libs. Thus
this patch removes flushing the cache added in port to 1.9.4.

(2) The from-file-loaded package poisons the cache. When querying
other packages, the from-file-loaded package could be included in the
results. I'm not sure whether the original isolation was good or bad.
I can imagine someone wants to override a system-provided pkgconfig
file and thus loads the override from a file.

Alternatively, we could cache the package loaded from a file
temporarily just around using the solver. That would narrow the time
window when the package is in the cache.

ppisar committed:
Handle a version in solver query

ppisar committed:
Use solver for list of cflags/libs

This resolves all the remaining tests.

ppisar committed:
Restore a trailing space in cflags and libs strings

pkgconf-1.9 fixed cflags and libs values by removing the trailing
spaces. However, this consitutes a change in behaviour (people got
used to concatenate flags) and thus this patch returns the trailing
spaces.

ppisar committed:
Make it buildable with pkgconf < 1.9 again

ppisar committed:
Cache packages loaded from file only temporarily

This implements an alternative approach outlined in
d3efe46b52b6ae3defb90cd695e835ebf6d13204 commit.

This prevents from having packages loaded from files in a client-wide
cache all the time. The cache is used for resolving cflags/libs.

A downside is that the code is more complex.

plicease commented:
Can you apply this patch:

```diff
diff --git a/.github/workflows/linux.yml b/.github/workflows/linux.yml
index c6ec3e8..f6cfb22 100644
--- a/.github/workflows/linux.yml
+++ b/.github/workflows/linux.yml
@@ -17,11 +17,12 @@ jobs:
       fail-fast: false
       matrix:
         cip:
-          - tag: "5.35"
-          - tag: "5.34"
-          - tag: "5.34"
+          - tag: "5.37"
+          - tag: "5.36"
+          - tag: "5.36"
             install_from_git: 1
-          - tag: "5.34-fedora34"
+          - tag: "5.36-fedora34"
+          - tag: "5.34"
           - tag: "5.32"
           - tag: "5.30"
           - tag: "5.28"
```
that should fix the CI failure.

ppisar committed:
CI: Update perl versions

ppisar commented:
I pushed the CI patch. 5.36,1 fails on a wrong autoconf version and 5.36-fedora34 fails on a missing docker image.  

Fedora 34 become unsupported a year ago. Now the oldest supported Fedora is 36. I recommend you instead of hard-coding a Fedora version, use fedora:latest for the latest stable Fedora, or fedora:rawhide for the latest developmental Fedora <https://hub.docker.com/_/fedora/tags>.

ppisar committed:
Fix resolving flags for packages with a name different from its identifier

Alien-Build-2.77 tests revealed a bug in constructing a query for
pkgconf-1.9 solver: If a package file had file name different from
a Name value inside the file, the package was able to be found, but
the flags solver searched for the Name and found nothing.

Studying pre-pkgconf documentation shows that Name value is only
a human-oriented display name and a base of the package file name
should be used instead as a package identifier. This base name is
stored into an id field of the package structure of pkgconf.

This patch fixes it by using the id field instead. It also adds a test
to prevent from future regressions.

ppisar commented:
Alien-Build tests found a mistake when retrieving flags for a package file whose Name field differs from its file name. At pushed a fix to the pull request.

felixonmars commented:
Hi. The Arch package has adopted this PR for compatibility with pkgconf 2.x, but unfortunately Alien-Build's test suite segfaults after this change. I have reported it to Alien-Build at https://github.com/PerlAlien/Alien-Build/issues/413 but since the real variable here is pkgconf and PkgConfig-LibPkgConf, I would like your help as well :)

felixonmars commented:
Oops. I figured it out. The problem here is pkgconf 2.1+ which probably brought more breakage and breaks PkgConfig-LibPkgConf's tests as well:

```
t/client.t .... 1/?
        #   Failed test at t/client.t line 219.
        #     Structures begin differing at:
        #          $got->[0] = '/tmp/Ew6Er88Ayt/foo'
        #     $expected->[0] = '/tmp/Ew6Er88Ayt/bar'
        # got:      [
        #   '/tmp/Ew6Er88Ayt/foo',
        #   '/tmp/Ew6Er88Ayt/bar',
        #   '/tmp/Ew6Er88Ayt/baz',
        #   '/tmp/Ew6Er88Ayt/ralph'
        # ]
        # expected: [
        #   '/tmp/Ew6Er88Ayt/bar',
        #   '/tmp/Ew6Er88Ayt/foo',
        #   '/tmp/Ew6Er88Ayt/ralph',
        #   '/tmp/Ew6Er88Ayt/baz'
        # ]

        #   Failed test at t/client.t line 222.
        #     Structures begin differing at:
        #          $got->[0] = '/tmp/Ew6Er88Ayt/trans'
        #     $expected->[0] = '/tmp/Ew6Er88Ayt/formers'
        # got:      [
        #   '/tmp/Ew6Er88Ayt/trans',
        #   '/tmp/Ew6Er88Ayt/formers'
        # ]
        # expected: [
        #   '/tmp/Ew6Er88Ayt/formers',
        #   '/tmp/Ew6Er88Ayt/trans'
        # ]

        #   Failed test at t/client.t line 225.
        #     Structures begin differing at:
        #          $got->[0] = '/tmp/Ew6Er88Ayt/trans'
        #     $expected->[0] = '/tmp/Ew6Er88Ayt/formers'
        # got:      [
        #   '/tmp/Ew6Er88Ayt/trans',
        #   '/tmp/Ew6Er88Ayt/formers'
        # ]
        # expected: [
        #   '/tmp/Ew6Er88Ayt/formers',
        #   '/tmp/Ew6Er88Ayt/trans'
        # ]
        # Looks like you failed 3 tests of 3.

    #   Failed test 'search path'
    #   at t/client.t line 227.

        #   Failed test at t/client.t line 235.
        #     Structures begin differing at:
        #          $got->[0] = '/tmp/Ew6Er88Ayt/foo/lib'
        #     $expected->[0] = '/tmp/Ew6Er88Ayt/bar/lib'
        # got:      [
        #   '/tmp/Ew6Er88Ayt/foo/lib',
        #   '/tmp/Ew6Er88Ayt/bar/lib'
        # ]
        # expected: [
        #   '/tmp/Ew6Er88Ayt/bar/lib',
        #   '/tmp/Ew6Er88Ayt/foo/lib'
        # ]

        #   Failed test at t/client.t line 238.
        #     Structures begin differing at:
        #          $got->[0] = '/tmp/Ew6Er88Ayt/trans/lib'
        #     $expected->[0] = '/tmp/Ew6Er88Ayt/formers/lib'
        # got:      [
        #   '/tmp/Ew6Er88Ayt/trans/lib',
        #   '/tmp/Ew6Er88Ayt/formers/lib'
        # ]
        # expected: [
        #   '/tmp/Ew6Er88Ayt/formers/lib',
        #   '/tmp/Ew6Er88Ayt/trans/lib'
        # ]
        # Looks like you failed 2 tests of 2.

    #   Failed test 'filter lib dirs'
    #   at t/client.t line 240.

        #   Failed test at t/client.t line 248.
        #     Structures begin differing at:
        #          $got->[0] = '/tmp/Ew6Er88Ayt/foo/include'
        #     $expected->[0] = '/tmp/Ew6Er88Ayt/bar/include'
        # got:      [
        #   '/tmp/Ew6Er88Ayt/foo/include',
        #   '/tmp/Ew6Er88Ayt/bar/include'
        # ]
        # expected: [
        #   '/tmp/Ew6Er88Ayt/bar/include',
        #   '/tmp/Ew6Er88Ayt/foo/include'
        # ]

        #   Failed test at t/client.t line 251.
        #     Structures begin differing at:
        #          $got->[0] = '/tmp/Ew6Er88Ayt/trans/include'
        #     $expected->[0] = '/tmp/Ew6Er88Ayt/formers/include'
        # got:      [
        #   '/tmp/Ew6Er88Ayt/trans/include',
        #   '/tmp/Ew6Er88Ayt/formers/include'
        # ]
        # expected: [
        #   '/tmp/Ew6Er88Ayt/formers/include',
        #   '/tmp/Ew6Er88Ayt/trans/include'
        # ]
        # Looks like you failed 2 tests of 2.

    #   Failed test 'filter include dirs'
    #   at t/client.t line 253.
    # Looks like you failed 3 tests of 3.

#   Failed test 'path attributes'
#   at t/client.t line 255.
# Looks like you failed 1 test of 13.
t/client.t .... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/13 subtests
```

Downgrading pkgconf to 2.0.3 makes the Alien-Build problem go away and the tests here pass too.

ppisar commented:
I didn't need to touch these patches since PkgConfig-LibPkgConf-0.11. I know that pkgconf recently broke ABI twice and leading to segfaults. A fix was rebuilding PkgConfig-LibPkgConf against the new pkgconf library.

ppisar commented:
Specifically the breakage was in pkgconf-2.1.0 and in pkgconf-2.1.1.

felixonmars commented:
Thanks for the info! Indeed a rebuild made the segfault go away. I have to ignore the test failures above though, probably that's the answer for now.

yselkowitz committed:
Adapt to pkgconf 2.5.0

pkgconf_pkg_new_from_file was replaced with pkgconf_pkg_new_from_path
which does not take a FILE pointer:

https://github.com/pkgconf/pkgconf/commit/61610c65cb7146cb6960ecbbe8902a26226fd794

This applies on top of "Adapt to pkgconf 1.9.4":

https://github.com/PerlAlien/PkgConfig-LibPkgConf/pull/17

ppisar committed:
CI: Update actions/cache to v5

v2 is blocked by GitHub.

fundawang commented:
Many thanks. But is it possible porting to pkgconf  3.0 also?
